src/cursor/abstract_cursor.ts

```diff
@@ -653,8 +653,15 @@ function next<T>(cursor: AbstractCursor, blocking: boolean, callback: Callback<T
 
   if (cursorId == null) {
     // All cursors must operate within a session, one must be made implicitly if not explicitly provided
-    if (cursor[kSession] == null && cursor[kTopology].hasSessionSupport()) {
-      cursor[kSession] = cursor[kTopology].startSession({ owner: cursor, explicit: false });
+    if (cursor[kSession] == null) {
+      if (cursor[kTopology].shouldCheckForSessionSupport()) {
+        return cursor[kTopology].selectServer(ReadPreference.primaryPreferred, err => {
+          if (err) return callback(err);
+          return next(cursor, blocking, callback);
+        });
+      } else if (cursor[kTopology].hasSessionSupport()) {
+        cursor[kSession] = cursor[kTopology].startSession({ owner: cursor, explicit: false });
+      }
     }
 
     cursor._initialize(cursor[kSession], (err, state) => {
```

Comment:
Is it true that cursor[kSession] will ever be null / undefined here? I noticed some spread logic in FindCursor that might be overriding a session from the options:

```ts
   const findOperation = new FindOperation(undefined, this.namespace, this[kFilter], {
      ...this[kBuiltOptions], // NOTE: order matters here, we may need to refine this
      ...this.cursorOptions,
      session
    });
```

I think if we are able to change this function to always take a defined session then we can be sure that this ^ isn't a bug.

Reply:
Good catch. With these changes we would never get to this code with an undefined session, so I've changed it both in  `FindCursor` and `AggregateCursor` to never accept `undefined`.

Reply:
This confuses me.  If the session is never undefined, why is the code in abstract_cursor that handles a null session still necessary?

Reply:
It handles both cases (null and undefined) as we are checking loose equality with `==`.